cmd/bosun/sched/check.go

```diff
@@ -200,58 +186,135 @@ func (s *Schedule) RunHistory(r *RunHistory) {
 	s.Save()
 }
 
-// CheckUnknown checks for unknown alerts.
-func (s *Schedule) CheckUnknown() {
-	for range time.Tick(s.Conf.CheckFrequency / 4) {
-		log.Println("checkUnknown")
-		r := s.NewRunHistory(time.Now(), nil)
-		s.Lock()
-		for ak, st := range s.status {
-			if st.Forgotten {
-				continue
-			}
-			a := s.Conf.Alerts[ak.Name()]
-			t := a.Unknown
-			if t == 0 {
-				t = s.Conf.CheckFrequency * 2
-			}
-			if t == 0 {
-				continue
-			}
-			if time.Since(st.Touched) < t {
-				continue
-			}
-			r.Events[ak] = &Event{Status: StUnknown}
+func (r *RunHistory) GetUnknownAndUnevaluatedAlertKeys(alert string) (unknown, uneval []expr.AlertKey) {
+	unknown = []expr.AlertKey{}
+	uneval = []expr.AlertKey{}
+	for ak, ev := range r.Events {
+		if ak.Name() != alert {
+			continue
+		}
+		if ev.Status == StUnknown {
+			unknown = append(unknown, ak)
+		} else if ev.Unevaluated {
+			uneval = append(uneval, ak)
+		}
+	}
+	return unknown, uneval
+}
+
+// Check evaluates all critical and warning alert rules. An error is returned if
+// the check could not be performed.
+func (s *Schedule) Check(T miniprofiler.Timer, now time.Time) (time.Duration, error) {
+	select {
+	case s.checkRunning <- true:
+		// Good, we've got the lock.
+	default:
+		return 0, fmt.Errorf("check already running")
+	}
+	r := s.NewRunHistory(now, cache.New(0))
+	start := time.Now()
+	for _, ak := range s.findUnknownAlerts(now) {
+		r.Events[ak] = &Event{Status: StUnknown}
+	}
+	for _, a := range s.Conf.Alerts {
+		s.CheckAlert(T, r, a)
+	}
+	d := time.Since(start)
+	s.RunHistory(r)
+	<-s.checkRunning
+	return d, nil
+}
+
+var bosunStartupTime = time.Now()
+
+func (s *Schedule) findUnknownAlerts(now time.Time) []expr.AlertKey {
+	keys := []expr.AlertKey{}
+	if time.Now().Sub(bosunStartupTime) < 1*time.Minute {
+		return keys
+	}
+	s.Lock()
+	for ak, st := range s.status {
+		if st.Forgotten {
+			continue
+		}
+		a := s.Conf.Alerts[ak.Name()]
+		t := a.Unknown
+		if t == 0 {
+			t = s.Conf.CheckFrequency * 2
 		}
-		s.Unlock()
-		s.RunHistory(r)
+		if now.Sub(st.Touched) < t {
+			continue
+		}
+		keys = append(keys, ak)
 	}
+	s.Unlock()
+	return keys
 }
 
 func (s *Schedule) CheckAlert(T miniprofiler.Timer, r *RunHistory, a *conf.Alert) {
 	log.Printf("check alert %v start", a.Name)
 	start := time.Now()
-	var warns expr.AlertKeys
-	crits, err := s.CheckExpr(T, r, a, a.Crit, StCritical, nil)
+	var warns, crits expr.AlertKeys
+	d, err := s.executeExpr(T, r, a, a.Depends)
+	var deps expr.ResultSlice
 	if err == nil {
-		warns, _ = s.CheckExpr(T, r, a, a.Warn, StWarning, crits)
+		deps = filterDependencyResults(d)
+		crits, err = s.CheckExpr(T, r, a, a.Crit, StCritical, nil)
+		if err == nil {
+			warns, _ = s.CheckExpr(T, r, a, a.Warn, StWarning, crits)
+		}
 	}
+	depCount, unknownCount := markDependenciesUnevaluated(r.Events, deps, a.Name)
```

Comment:
That is pretty silly to set up, and I would call it a mistake anytime I saw it in real life. That alert has two real states: Normal and unevaluated. It may go crit/unevaluated, warn/unevaluated or unknown/unevaluated, but none of those will notify or show up on the dashboard.

Short answer: don't do that.